kv/kvserver: TestStrictGCEnforcement failed

kv/kvserver.TestStrictGCEnforcement [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3791324&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3791324&tab=artifacts#/) on master @ [0331536f74eb8a2bf5b2ce20466c9786e37c1ac9](https://github.com/cockroachdb/cockroach/commits/0331536f74eb8a2bf5b2ce20466c9786e37c1ac9):


```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement3372521210
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3334: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement3372521210
--- FAIL: TestStrictGCEnforcement (1.89s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3332: 
        	Error Trace:	client_replica_test.go:3201
        	            				client_replica_test.go:3332
        	Error:      	Received unexpected error:
        	            	batch timestamp 1637774509.637511107,0 must be after replica GC threshold 1637774518.771840542,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.04s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStrictGCEnforcement.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStrictGCEnforcement [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3792356&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3792356&tab=artifacts#/) on master @ [0331536f74eb8a2bf5b2ce20466c9786e37c1ac9](https://github.com/cockroachdb/cockroach/commits/0331536f74eb8a2bf5b2ce20466c9786e37c1ac9):


```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement479635463
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3334: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement479635463
--- FAIL: TestStrictGCEnforcement (1.49s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3332: 
        	Error Trace:	client_replica_test.go:3201
        	            				client_replica_test.go:3332
        	Error:      	Received unexpected error:
        	            	batch timestamp 1637781428.728210662,0 must be after replica GC threshold 1637781437.825271906,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.04s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStrictGCEnforcement.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -3142,7 +3142,6 @@ func TestTransferLeaseBlocksWrites(t *testing.T) {
 // to the zone configs.
 func TestStrictGCEnforcement(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 73123, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	// The unfortunate thing about this test is that the gcttl is in seconds and

--- a/pkg/kv/kvserver/replica.go
+++ b/pkg/kv/kvserver/replica.go
@@ -1384,9 +1384,8 @@ func (r *Replica) checkGCThresholdAndLeaseRLocked(
 	//
 	// If the request is an INCONSISTENT request (and thus a read), it similarly
 	// doesn't check the lease.
-	leaseChecked := !ba.IsSingleSkipsLeaseCheckRequest() && ba.ReadConsistency != roachpb.INCONSISTENT
-	if leaseChecked {
-		// Now check the lease.
+	if !ba.IsSingleSkipsLeaseCheckRequest() && ba.ReadConsistency != roachpb.INCONSISTENT {
+		// Check the lease.
 		var err error
 		shouldExtend, err = r.leaseGoodToGoForStatusRLocked(ctx, now, reqTS, st)
 		if err != nil {
@@ -1397,8 +1396,12 @@ func (r *Replica) checkGCThresholdAndLeaseRLocked(
 				return kvserverpb.LeaseStatus{}, false, err
 			}
 			// Otherwise, suppress the error. Also, remember that we're not serving
-			// this under the lease.
-			leaseChecked, err = false, nil
+			// this under the lease by zeroing out the status. We also intentionally
+			// do not pass the original status to checkTSAboveGCThresholdRLocked as
+			// this method assumes that a valid status indicates that this replica
+			// holds the lease (see #73123). `shouldExtend` is already false in this
+			// branch, but for completeness we zero it out as well.
+			st, shouldExtend, err = kvserverpb.LeaseStatus{}, false, nil
 		}
 	}
 
@@ -1409,9 +1412,6 @@ func (r *Replica) checkGCThresholdAndLeaseRLocked(
 		return kvserverpb.LeaseStatus{}, false, err
 	}
 
-	if !leaseChecked {
-		return kvserverpb.LeaseStatus{}, false, nil
-	}
 	return st, shouldExtend, nil
 }
 

